File: 30_2.hs
{--
Задача 30-2
Одеров Роман, 545 гр.
--}

{--
Задача о поиске часто встречающихся элементов.

В общем случае можно реализовать поиск элементов,
встречающихся более N/k раз, за O(k) памяти и O(n).

см. 
http://habrahabr.ru/post/167177/
http://keithschwarz.com/interesting/code/?dir=majority-element
--}

findMajor [] = Nothing
findMajor xs = 	let
					temp = fm xs 0 0
				in
					if check xs temp 0 0
					then Just temp
					else Nothing


fm [] stand candidate = candidate
fm (x:xs) stand candidate =
	if stand == 0
	then fm xs (stand+1) x
	else
		if candidate == x
		then fm xs (stand+1) candidate
		else fm xs (stand-1) candidate


check xs t count_t len_xs
	|xs == [] = (count_t > (len_xs /2))
	|otherwise =if head xs == t
				then check (tail xs) t (count_t+1) (len_xs+1)
				else check (tail xs) t (count_t) (len_xs+1)

File: 30_2.c
int* majority(int[] array, int N) {
  int confidence = 0; // количество людей, не нашедших пары и оставшихся стоять
  int* candidate = NULL; // последний человек, не нашедший пару -- 
                         // возможно, его элемент встречается чаще всего

  // проходим по массиву и усаживаем пары с разными элементами
  for (int i = 0; i<N; i++) {

    // если до сих пор все сидят, то следующему пока придётся постоять
    if (confidence == 0) {
      candidate = array+i;
      confidence++;
    }

    // иначе следующий либо садится с одним из стоящих,
    // либо будет стоять вместе с ними, если у него такой же элемент
    else if (*candidate == array[i])) 
      confidence++;
    else 
      confidence--;
  }

  return confidence > 0 ? candidate : NULL;
}